File: ios/Plugin/MarketingCloudPlugin.swift
import Foundation
import Capacitor
import MarketingCloudSDK

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(MarketingCloudPlugin)
public class MarketingCloudPlugin: CAPPlugin {
    private let implementation = MarketingCloud()

    override public func load() {
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(self.handleNotificationOpened(notification:)),
                                               name: NSNotification.Name.SFMCFoundationUNNotificationReceived,
                                               object: nil)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    @objc public func isPushEnabled(_ call: CAPPluginCall) {
        let isEnabled = implementation.isPushEnabled()
        call.resolve([ "value": isEnabled ])
    }

    @objc public func setPushEnabled(_ call: CAPPluginCall) {
        guard let enabled = call.getBool("enabled") else {
            return call.reject("Error enabling push since the token is null")
        }
        implementation.setPushEnabled(enabled)
    }

    @objc public func getProfileId(_ call: CAPPluginCall) {
        call.resolve([ "value": implementation.getProfileId() ?? NSNull() ])
    }

    @objc public func setProfileId(_ call: CAPPluginCall) {
        guard let id = call.getString("value") else {
            return call.reject("Error setting profile id since its value is null")
        }
        implementation.setProfileId(id)
    }

    @objc public func getAttributes(_ call: CAPPluginCall) {
        call.resolve([ "attributes": implementation.getAttributes() ?? [:] ])
    }

    @objc public func setAttribute(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), let value = call.getString("value") else {
            return call.reject("Invalid attribute key or value")
        }
        implementation.setAttribute(key, value)
    }

    @objc public func clearAttribute(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            return call.reject("Invalid attribute key")
        }
        call.resolve([ "value": implementation.clearAttribute(key) ])
    }

    @objc public func getTags(_ call: CAPPluginCall) {
        call.resolve([ "tags": implementation.getTags() ?? [] ])
    }

    @objc public func addTag(_ call: CAPPluginCall) {
        guard let tag = call.getString("value") else {
            return call.reject("Invalid tag")
        }
        call.resolve([ "value": implementation.addTag(tag) ])
    }

    @objc public func removeTag(_ call: CAPPluginCall) {
        guard let tag = call.getString("value") else {
            return call.reject("Invalid tag")
        }
        call.resolve([ "value": implementation.removeTag(tag) ])
    }

    @objc public func isMarketingCloudNotification(_ call: CAPPluginCall) {
        let notification = call.getObject("notification", [:])
        let isSfmcPush = implementation.isMarketingCloudNotification(notification)
        call.resolve([ "value": isSfmcPush ])
    }

    @objc public func notifyNotificationOpened(_ call: CAPPluginCall) {
        let notification = call.getObject("notification", [:])
        implementation.notifyNotificationOpened(notification)
    }

    @objc public func handleNotificationOpened(notification: NSNotification) {
        guard var notificationMessage = MarketingCloud.extractNotificationMessage(notification) else {
            return;
        }

        let messageId = MarketingCloudNotification.getId(notificationMessage)
        if messageId != nil {
            notificationMessage["sfmcMessageId"] = messageId
        }

        func addUrl(_ url: String) { notificationMessage["url"] = url }

        func addType(_ type: String) {notificationMessage["type"] = type }

        if let openDirectUrl = MarketingCloudNotification.getOpenDirectUrl(notificationMessage) {
            addUrl(openDirectUrl)
            addType("openDirect")
        } else if let cloudPageUrl = MarketingCloudNotification.getCloudPageUrl(notificationMessage)  {
            addUrl(cloudPageUrl)
            addType("cloudPage")
        } else {
            addType("other")
        }

        notifyListeners("notificationOpened", data: [
            "sfmcMessageId": messageId ?? NSNull(),
            "extras": notificationMessage,
            "action": "tap"
        ])
    }

}

File: android/src/main/java/com/okode/marketingcloud/MarketingCloudPlugin.kt
package com.okode.marketingcloud

import android.content.Intent
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.salesforce.marketingcloud.notifications.NotificationMessage
import org.json.JSONObject

@CapacitorPlugin(name = "MarketingCloud")
class MarketingCloudPlugin : Plugin() {

    private val implementation = MarketingCloud()

    override fun handleOnNewIntent(intent: Intent?) {
        super.handleOnNewIntent(intent)
        handleNotificationOpened(MarketingCloudSdkConfig.extractNotificationMessage(intent))
    }

    @PluginMethod
    fun isPushEnabled(call: PluginCall) {
        implementation.isPushEnabled {
            call.resolve(JSObject().put("value", it))
        }
    }

    @PluginMethod
    fun setPushEnabled(call: PluginCall) {
        val enabled = call.getBoolean("enabled", true)!!
        implementation.setPushEnabled(enabled)
    }

    @PluginMethod
    fun getProfileId(call: PluginCall) {
        implementation.getProfileId() {
            call.resolve(JSObject().put("value", it))
        }
    }

    @PluginMethod
    fun setProfileId(call: PluginCall) {
        val profileId = call.getString("value")
        if (profileId == null) {
            call.reject("Error setting profile id since its value is null")
            return
        }
        implementation.setProfileId(profileId)
    }

    @PluginMethod
    fun getAttributes(call: PluginCall) {
        implementation.getAttributes { it
            call.resolve(JSObject().put("attributes", it))
        }
    }

    @PluginMethod
    fun setAttribute(call: PluginCall) {
        val key = call.getString("key")
        val value = call.getString("value")
        if (key == null || value == null) {
            call.reject("Invalid attribute key or value")
            return
        }
        implementation.setAttribute(key, value)
    }

    @PluginMethod
    fun clearAttribute(call: PluginCall) {
        val key = call.getString("key")
        if (key == null) {
            call.reject("Invalid attribute key")
            return
        }
        implementation.clearAttribute(key) {
            call.resolve(JSObject().put("value", it))
        }
    }

    @PluginMethod
    fun getTags(call: PluginCall) {
        implementation.getTags { it
            call.resolve(JSObject().put("tags", it))
        }
    }

    @PluginMethod
    fun addTag(call: PluginCall) {
        val tag = call.getString("value")
        if (tag == null) {
            call.reject("Invalid tag")
            return
        }
        implementation.addTag(tag) {
            call.resolve(JSObject().put("value", it))
        }
    }

    @PluginMethod
    fun removeTag(call: PluginCall) {
        val tag = call.getString("value")
        if (tag == null) {
            call.reject("Invalid tag")
            return
        }
        implementation.removeTag(tag) {
            call.resolve(JSObject().put("value", it))
        }
    }

    @PluginMethod
    fun isMarketingCloudNotification(call: PluginCall) {
        val notification = call.getObject("notification", JSObject())!!
        val result = MarketingCloud.isMarketingCloudNotification(notification)
        call.resolve(JSObject().put("value", result))
    }

    @PluginMethod
    fun showNotification(call: PluginCall) {
        val notification = call.getObject("notification", JSObject())!!
        MarketingCloud.showNotification(notification) {
            call.resolve(JSObject().put("value", it))
        }
    }

    private fun handleNotificationOpened(notification: NotificationMessage?) {
        if (notification == null) { return }
        val values = JSObject.fromJSONObject(JSONObject(notification.payload))
        val messageId = notification.payload?.get("_m")
        values.put("sfmcMessageId", messageId)
        notification.url?.let { values.put("url", it) }
        val notificationType = when(notification.type) {
            NotificationMessage.Type.OTHER -> "other"
            NotificationMessage.Type.CLOUD_PAGE -> "cloudPage"
            NotificationMessage.Type.OPEN_DIRECT -> "openDirect"
        }
        values.put("type", notificationType)
        val event = JSObject()
                .put("sfmcMessageId", messageId)
                .put("extras", values)
                .put("action", "tap")
        notifyListeners("notificationOpened", event, true)
    }

}